File: macos/Sources/ElevateCore/Coordination/AccessPackageDiff.swift
import Foundation

/// Something worth telling the user about between two polls of a tenant's access packages.
public enum AccessPackageEvent: Hashable, Sendable {
    case approved(AccessPackageRequest)
    case denied(AccessPackageRequest)
    case deliveryFailed(AccessPackageRequest)
    case revoked(AccessPackageAssignment)
    case expired(AccessPackageAssignment)

    public var packageName: String {
        switch self {
        case .approved(let r), .denied(let r), .deliveryFailed(let r): r.packageName
        case .revoked(let a), .expired(let a): a.packageName
        }
    }
}

/// Pure comparison of two snapshots. A nil `previous` is the first sight of a tenant and never
/// notifies: everything present then is baseline, not news.
public enum AccessPackageDiff {
    public static func events(previous: AccessPackageSnapshot?, current: AccessPackageSnapshot, now: Date) -> [AccessPackageEvent] {
        guard let previous else { return [] }
        var out: [AccessPackageEvent] = []

        // Requests: only a request we already knew in an open state can transition into news.
        let previousStates = Dictionary(previous.requests.map { ($0.id, $0.state) }, uniquingKeysWith: { a, _ in a })
        for r in current.requests {
            guard let was = previousStates[r.id], was.isOpen, was != r.state else { continue }
            switch r.state {
            case .delivered: out.append(.approved(r))
            case .denied: out.append(.denied(r))
            case .deliveryFailed: out.append(.deliveryFailed(r))
            default: break
            }
        }

        // Assignments: a delivered one that is gone or no longer delivered is either revoked or
        // expired. Graph's own "expired" state wins; otherwise the end date decides, so an
        // assignment that vanished after its end date reads as expired, not revoked.
        let currentById = Dictionary(current.assignments.map { ($0.id, $0) }, uniquingKeysWith: { a, _ in a })
        for a in previous.assignments where a.state == .delivered {
            let still = currentById[a.id]?.state
            guard still != .delivered else { continue }
            if still == .expired {
                out.append(.expired(a))
            } else if let end = a.expiresAt, end <= now {
                out.append(.expired(a))
            } else {
                out.append(.revoked(a))
            }
        }
        return out
    }
}

File: windows/src/Elevate.Core/Coordination/AccessPackageDiff.cs
using Elevate.Core.Models;

namespace Elevate.Core.Coordination;

/// <summary>Something worth telling the user about between two polls of a tenant's access packages.</summary>
public abstract record AccessPackageEvent
{
    private AccessPackageEvent()
    {
    }

    public abstract string PackageName { get; }

    public sealed record Approved(AccessPackageRequest Request) : AccessPackageEvent
    {
        public override string PackageName => Request.PackageName;
    }

    public sealed record Denied(AccessPackageRequest Request) : AccessPackageEvent
    {
        public override string PackageName => Request.PackageName;
    }

    public sealed record DeliveryFailed(AccessPackageRequest Request) : AccessPackageEvent
    {
        public override string PackageName => Request.PackageName;
    }

    public sealed record Revoked(AccessPackageAssignment Assignment) : AccessPackageEvent
    {
        public override string PackageName => Assignment.PackageName;
    }

    public sealed record Expired(AccessPackageAssignment Assignment) : AccessPackageEvent
    {
        public override string PackageName => Assignment.PackageName;
    }
}

/// <summary>
/// Pure comparison of two snapshots. A null <c>previous</c> is the first sight of a tenant and
/// never notifies: everything present then is baseline, not news. Port of the Swift <c>AccessPackageDiff</c>.
/// </summary>
public static class AccessPackageDiff
{
    public static IReadOnlyList<AccessPackageEvent> Events(AccessPackageSnapshot? previous, AccessPackageSnapshot current, DateTimeOffset now)
    {
        ArgumentNullException.ThrowIfNull(current);
        if (previous is null)
        {
            return [];
        }

        var events = new List<AccessPackageEvent>();

        // Requests: only a request we already knew in an open state can transition into news.
        var previousStates = new Dictionary<string, AccessPackageRequestState>(StringComparer.Ordinal);
        foreach (var r in previous.Requests)
        {
            previousStates.TryAdd(r.Id, r.State);
        }

        foreach (var r in current.Requests)
        {
            if (!previousStates.TryGetValue(r.Id, out var was) || !was.IsOpen() || was == r.State)
            {
                continue;
            }

            switch (r.State)
            {
                case AccessPackageRequestState.Delivered:
                    events.Add(new AccessPackageEvent.Approved(r));
                    break;
                case AccessPackageRequestState.Denied:
                    events.Add(new AccessPackageEvent.Denied(r));
                    break;
                case AccessPackageRequestState.DeliveryFailed:
                    events.Add(new AccessPackageEvent.DeliveryFailed(r));
                    break;
                default:
                    break;
            }
        }

        // Assignments: a delivered one that is gone or expired is either revoked or expired,
        // decided by whether its end date has passed.
        var currentById = new Dictionary<string, AccessPackageAssignment>(StringComparer.Ordinal);
        foreach (var a in current.Assignments)
        {
            currentById.TryAdd(a.Id, a);
        }

        foreach (var a in previous.Assignments.Where(a => a.State == AccessPackageAssignmentState.Delivered))
        {
            if (currentById.TryGetValue(a.Id, out var still) && still.State == AccessPackageAssignmentState.Delivered)
            {
                continue;
            }

            events.Add(a.ExpiresAt is { } end && end <= now
                ? new AccessPackageEvent.Expired(a)
                : new AccessPackageEvent.Revoked(a));
        }

        return events;
    }
}
